In [1]:
# Python 3.10.15

# Import necessary libraries
import os
import numpy as np
import mne
import matplotlib.pyplot as plt
import seaborn as sns

# TensorFlow and Keras 2.15.0
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras.layers import (Input, Conv2D, BatchNormalization, Activation, MaxPooling2D, Dropout, Flatten, Dense)
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

# Scikit-learn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, classification_report

# Suppress TensorFlow warnings (optional)
import logging
logging.getLogger('tensorflow').setLevel(logging.ERROR)


In [2]:
# Data Preparation
# Define the data directory where subject folders are located
data_dir = '/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed'

# List all subject folders
subject_folders = [sub for sub in os.listdir(data_dir) if os.path.isdir(os.path.join(data_dir, sub))]
'''
subjects = []
for sub in os.listdir(data_dir):
    if os.path.isdir(os.path.join(data_dir, sub)):
        subjects.append(sub)
'''

# Initialize lists to hold data and labels from all subjects
X_list = []
y_list = []

# Loop over each subject folder
for subject in subject_folders:
    
    subject_data_dir = os.path.join(data_dir, subject)
    
    # Construct file paths for the subject's data
    X_file = os.path.join(subject_data_dir, f'Epochs_{subject}.fif')
    
    # Check if data files exist
    if os.path.exists(X_file):
        
        print()
        print(f'***** Loading the processed data: {subject}')
        print()
        
        # Load the data
        X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
        y_subject = X_subject.events[:, 2]
        
        # Append to the list
        X_sub_data = X_subject.get_data()
        X_list.append(X_sub_data)
        y_list.append(y_subject)
        
    else:
        print()
        print(f'***** Data file Does Not Exist: {subject}')
        print()

# Ensure that at least one subject has been loaded
if len(X_list) == 0:
    print()
    raise ValueError("No data was loaded. Please check your data directory and files.")



***** Loading the processed data: sub-021



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-021/Epochs_sub-021.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-026


***** Loading the processed data: sub-019



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-026/Epochs_sub-026.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-019/Epochs_sub-019.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-010


***** Loading the processed data: sub-017



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-010/Epochs_sub-010.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-017/Epochs_sub-017.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-028


***** Loading the processed data: sub-016


***** Loading the processed data: sub-029



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-028/Epochs_sub-028.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-016/Epochs_sub-016.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-029/Epochs_sub-029.fif) does not conform to MNE naming conventi


***** Loading the processed data: sub-011


***** Loading the processed data: sub-027



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-011/Epochs_sub-011.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-027/Epochs_sub-027.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-018


***** Loading the processed data: sub-020



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-018/Epochs_sub-018.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-020/Epochs_sub-020.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-002


***** Loading the processed data: sub-005



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-002/Epochs_sub-002.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-005/Epochs_sub-005.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-033


***** Loading the processed data: sub-034



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-033/Epochs_sub-033.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-034/Epochs_sub-034.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-035


***** Loading the processed data: sub-032


***** Loading the processed data: sub-004



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-035/Epochs_sub-035.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-032/Epochs_sub-032.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-004/Epochs_sub-004.fif) does not conform to MNE naming conventi


***** Loading the processed data: sub-003


***** Loading the processed data: sub-040



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-003/Epochs_sub-003.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-040/Epochs_sub-040.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-025


***** Loading the processed data: sub-022



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-025/Epochs_sub-025.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-022/Epochs_sub-022.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-014


***** Loading the processed data: sub-013



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-014/Epochs_sub-014.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-013/Epochs_sub-013.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-012


***** Loading the processed data: sub-015



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-012/Epochs_sub-012.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-015/Epochs_sub-015.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-023


***** Loading the processed data: sub-024



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-023/Epochs_sub-023.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-024/Epochs_sub-024.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-039


***** Loading the processed data: sub-006



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-039/Epochs_sub-039.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-006/Epochs_sub-006.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-001


***** Loading the processed data: sub-008


***** Loading the processed data: sub-037



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-001/Epochs_sub-001.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-008/Epochs_sub-008.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-037/Epochs_sub-037.fif) does not conform to MNE naming conventi


***** Loading the processed data: sub-030


***** Loading the processed data: sub-031



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-030/Epochs_sub-030.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-031/Epochs_sub-031.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-009


***** Loading the processed data: sub-036



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-009/Epochs_sub-009.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-036/Epochs_sub-036.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)



***** Loading the processed data: sub-038


***** Loading the processed data: sub-007



/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-038/Epochs_sub-038.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)
/var/folders/dh/551p960n6mv9t9byzj8pp0640000gp/T/ipykernel_4708/3738345600.py:34: RuntimeWarning: This filename (/Users/BAEK/Code/neurEx/data/N170/Data_Preprocessed/sub-007/Epochs_sub-007.fif) does not conform to MNE naming conventions. All epochs files should end with -epo.fif, -epo.fif.gz, _epo.fif or _epo.fif.gz
  X_subject = mne.read_epochs(X_file, preload=True, verbose=False)


### P8 (Right Parietal): 

The parietal lobe is involved in sensory processing and the integration of sensory information. Specifically, the right parietal region is important for spatial awareness, visual processing, and attention. It is often engaged in tasks related to visual perception and sensory-motor coordination.

### PO8 (Right Parieto-Occipital): 

The occipital lobe is the primary visual processing center in the brain, and the parietal lobe integrates sensory information. The PO8 region is crucial for processing visual information from both the environment and sensory input. It is especially important in the recognition of objects, including faces, and is involved in spatial processing and attention to visual stimuli.

### O2 (Right Occipital): 

The occipital lobe is the brain’s main area for visual processing, including perception of visual stimuli such as shapes, colors, and faces. The right occipital lobe is particularly active during tasks related to visual processing and recognition of visual patterns.

### P10 (Right Parietal-Temporal): 

The parietal lobe is involved in sensory integration, spatial awareness, and attention, while the temporal lobe is important for processing sensory input, especially auditory and visual information. The temporal lobe is heavily involved in memory, recognition, and face processing.


        a = np.array([[1, 2], 
                    [3, 4]])
        b = np.array([[5, 6], 
                    [7, 8]])

        np.concatenate((a, b), axis=0)
        # [[1, 2],
        #  [3, 4],
        #  [5, 6],
        #  [7, 8]]

        np.vstack((a,b))
        # [[1, 2],
        #  [3, 4],
        #  [5, 6],
        #  [7, 8]]

        np.concatenate((a, b), axis=1)
        # [[1 2 5 6]
        #  [3 4 7 8]]

        a = np.array([[[1, 2], 
                    [3, 4]]])  # Shape: (1, 2, 2)

        b = np.array([[[5, 6], 
                    [7, 8]]])  # Shape: (1, 2, 2)

        np.concatenate((a, b), axis=2)
        # [[[1, 2, 5, 6],
        #   [3, 4, 7, 8]]]

        np.vstack((a,b))
        # [[[1, 2],
        #   [3, 4]],
        #
        #   [[5, 6],
        #    [7, 8]]]





        # Mock EEG data with shape (3 samples, 4 channels, 5 timepoints)
        X = np.array([
            [[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15], [16, 17, 18, 19, 20]],
            [[21, 22, 23, 24, 25], [26, 27, 28, 29, 30], [31, 32, 33, 34, 35], [36, 37, 38, 39, 40]],
            [[41, 42, 43, 44, 45], [46, 47, 48, 49, 50], [51, 52, 53, 54, 55], [56, 57, 58, 59, 60]]
        ])
        # Shape: (3 samples, 4 channels, 5 timepoints)

        chan_idx = [1, 3]  # Select only channels 1 and 3

        X = X[:, chan_idx, :]

        # Extracts only the specified channels
        X = np.array([
            [[6, 7, 8, 9, 10], [16, 17, 18, 19, 20]],
            [[26, 27, 28, 29, 30], [36, 37, 38, 39, 40]],
            [[46, 47, 48, 49, 50], [56, 57, 58, 59, 60]]
        ])
        # Shape: (3 samples, 2 channels, 5 timepoints)





        # Labels for three subjects
        y_list = [
            np.array([0, 1, 0, 1]),  # Subject 1
            np.array([1, 1, 0, 0]),  # Subject 2
            np.array([0, 0, 1, 1])   # Subject 3
        ]

        np.concatenate(y_list, axis=0)
        #[0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1])  # Shape: (12,)

In [ ]:
### Get channel names from the last loaded subject
# This retrieves the names of the channels from the EEG data. These names represent the locations or positions of the EEG electrodes on the scalp, like ‘FP1’, ‘F3’, ‘P3’, etc.
channels = X_subject.ch_names

# Channels such as P8, PO8, O2, and P10 are situated over or near these regions in the parietal and occipital lobes, which are strongly involved in facial processing and visual stimuli processing.
# Studies on the N170 often report that the most reliable and strongest N170 signals are found in right-lateralized occipital-temporal regions. 
coi = ['P8', 'PO8', 'O2', 'P10']

# Creates a list of indices (chan_idx) for the channels of interest. 
# It finds the index (position) of each channel in the coi list from the full channels list.
chan_idx = [channels.index(chan) for chan in coi] # [25, 27, 29, 26]
'''
chan_idx = []
for chan in coi:
    index = channels.index(chan)    # Get the index of the channel in the full list
    chan_idx.append(index)          # Append the index to the list
'''

### Concatenate data from all subjects
# X_list: This is a list containing the EEG data arrays for multiple subjects. Each array in X_list typically has a shape like (trials, channels, time_points).

# NumPy arrays are multi-dimensional, and each dimension is associated with an axis:
# 0 represents the rows (vertical direction).
# 1 represents the columns (horizontal direction).
# 2, 3, and beyond represent additional dimensions (for higher-dimensional arrays).

# Combines the EEG data for all subjects along the trials dimension. Same as X = np.vstack(X_list) when concatenate(X_list, axis = 0)
# After concatenation, X will contain all trials from all subjects in a single array.
# If, for example, each subject’s data has 100 trials, and there are 10 subjects, then X will have a shape (1000, channels, time_points).
X = np.concatenate(X_list, axis = 0) # will combine all x of X(x,y,z)

# A slicing operation on the X array, and it extracts specific subsets of the data along its second axis (dimensions).
X = X[:, chan_idx, :]  # will change the y of X(x,y,z)

# Combines all the individual y arrays in y_list into a single, larger 1D or 2D array along the first axis (axis=0)
y = np.concatenate(y_list, axis = 0)

print(f'Combined data shape before filtering: X={X.shape}, y={y.shape}')

### Filter stimulus events
stimulus_labels = [1, 2]  # 1: Face, 2: Car
# Checks each element in y to see if it is in the stimulus_labels list.
# Returns a boolean array (stimulus_mask) of the same length as y.
stimulus_mask = np.isin(y, stimulus_labels)
X = X[stimulus_mask] # will change the x of X(x,y,z)
y = y[stimulus_mask]

# Adjust labels to start from 0
y = y - 1

# Focus on the N170 response by selecting data around 170 ms 
# Selecting Data Around the N170 Response (time window selection)
# This defines the time window you’re interested in, i.e., 100ms to 200ms after the stimulus.
# This means you are interested in data collected from 0.1 seconds to 0.2 seconds after the stimulus
tmin, tmax = 0.10, 0.2   # 100 ms to 200 ms
times = X_subject.times  # Time vector from the epochs
# Filter the times so that only the times that lies between tmin and tmax stay
time_mask = (times >= tmin) & (times <= tmax)
X_focused = X[:, :, time_mask]

print(f'Combined data shape after filtering: X={X_focused.shape}, y={y.shape}')

X_focused[0,0]

Combined data shape before filtering: X=(23602, 4, 820), y=(23602,)
Combined data shape after filtering: X=(6104, 4, 102), y=(6104,)


array([-2.65535473e-07, -2.33084976e-07, -2.56221681e-07, -3.35722922e-07,
       -4.71088408e-07, -6.60497783e-07, -9.00880692e-07, -1.18786204e-06,
       -1.51586949e-06, -1.87812459e-06, -2.26676702e-06, -2.67311763e-06,
       -3.08767533e-06, -3.50033306e-06, -3.90088247e-06, -4.27891778e-06,
       -4.62421035e-06, -4.92699383e-06, -5.17809914e-06, -5.36951111e-06,
       -5.49421166e-06, -5.54619740e-06, -5.52102088e-06, -5.41581248e-06,
       -5.22933673e-06, -4.96202801e-06, -4.61585473e-06, -4.19435119e-06,
       -3.70253657e-06, -3.14671301e-06, -2.53441977e-06, -1.87421798e-06,
       -1.17549646e-06, -4.48325484e-07,  2.96874105e-07,  1.04961514e-06,
        1.79936683e-06,  2.53584337e-06,  3.24924111e-06,  3.93041848e-06,
        4.57091855e-06,  5.16319560e-06,  5.70080136e-06,  6.17847393e-06,
        6.59201477e-06,  6.93824957e-06,  7.21520756e-06,  7.42191790e-06,
        7.55842064e-06,  7.62581203e-06,  7.62584541e-06,  7.56124113e-06,
        7.43529985e-06,  

In [6]:
print(X_focused.shape)

X_focused

(6104, 4, 102)


array([[[-2.65535473e-07, -2.33084976e-07, -2.56221681e-07, ...,
          2.21065354e-06,  2.27832317e-06,  2.31764697e-06],
        [ 1.52646923e-06,  1.55423677e-06,  1.55282390e-06, ...,
          7.41099594e-07,  9.07904801e-07,  1.05732774e-06],
        [-7.22147654e-06, -7.38670919e-06, -7.53458689e-06, ...,
         -4.07241284e-07, -2.46243744e-07, -8.30428524e-08],
        [ 5.03884314e-06,  5.25305508e-06,  5.36733101e-06, ...,
          8.15229414e-06,  8.16074655e-06,  8.08678148e-06]],

       [[-1.47737646e-06, -1.67229044e-06, -1.82316827e-06, ...,
          3.79423474e-06,  3.87306189e-06,  3.89641451e-06],
        [-6.68424724e-07, -6.97640298e-07, -7.08379923e-07, ...,
         -1.09889269e-06, -8.65716336e-07, -6.51313899e-07],
        [-7.48422084e-07, -9.07281396e-07, -1.03889632e-06, ...,
         -8.53871045e-07, -6.54174386e-07, -4.62437718e-07],
        [-4.21238231e-06, -4.83159445e-06, -5.40723513e-06, ...,
         -1.00315418e-08,  2.54398673e-07,  5.16855

1. X

NumPy arrays are multidimensional data structures. Each dimension corresponds to a different aspect of the data.

First Dimension (595):
•	Represents 595 epochs (or trials) in your experiment. Each epoch corresponds to a single trial where the EEG data is collected during a stimulus presentation or event.
	•	This dimension will vary as you add more subjects or trials, but each epoch will contain data in the format specified by the next two dimensions.

Second Dimension (33):
•	Represents 33 EEG channels (electrodes) capturing data from various scalp locations. Each channel corresponds to a different electrode that records electrical activity from the brain.
•	These 33 channels provide spatial information about where the brain activity is occurring on the scalp.
•	This dimension will remain the same across all subjects, as you are using the same number of electrodes to capture the data.

Third Dimension (820):
•	Represents 820 time points for each epoch. These are the EEG values recorded over time during each trial. The EEG signal is continuous, and the time points correspond to the sampling rate and the duration of the recording.
•	This dimension also remains the same across all subjects, as the data is sampled at the same rate and duration for each trial.

2. For example 

import numpy as np
box = np.random.rand(10, 20, 5)  # Generate random values in a (10, 20, 5) shape
print(box)

  array([[[0.72, 0.43, 0.99, 0.58, 0.16],  # Epoch 1, Channel 1
          [0.63, 0.34, 0.88, 0.44, 0.01],  # Epoch 1, Channel 2
          ...
          [0.56, 0.11, 0.98, 0.75, 0.22]], # Epoch 1, Channel 20
         [[0.89, 0.22, 0.78, 0.33, 0.55],  # Epoch 2, Channel 1
          [0.71, 0.48, 0.64, 0.88, 0.19],
          ...
          [0.12, 0.95, 0.72, 0.66, 0.23]], # Epoch 2, Channel 20
          ...
         [[0.82, 0.73, 0.61, 0.88, 0.34],  # Epoch 10, Channel 1
          [0.44, 0.56, 0.77, 0.93, 0.18],
          ...
          [0.62, 0.29, 0.38, 0.99, 0.41]]  # Epoch 10, Channel 20
        ])

Step 1: Mapping xx to Excel Files

The structure of your EEG data (xx) can be thought of as:
	•	Rows in the Excel sheet (y-axis): These correspond to the 33 EEG channels.
	•	Columns in the Excel sheet (x-axis): These correspond to the 820 time points.
	•	Multiple Excel sheets (z-axis): Each sheet corresponds to one epoch (595 epochs total).

Imagine one epoch of data as a single Excel file. For Epoch 1, you have a 33×820 table:

Channel/Time	 t1	    t2	 t3	  …	t820
         Ch1	0.12	-0.34	0.56	…	0.44
         Ch2	0.15	-0.23	0.54	…	0.39
         Ch3	0.10	-0.18	0.45	…	0.35
          …	   …	    …	   …	  …	 …
         Ch33	0.22	-0.28	0.67	…	0.50

Step 3: Visualizing Across Epochs

Now, you stack 595 Excel files one on top of the other. Each file has the same dimensions (33×820), but the data values inside them differ because each epoch captures a slightly different response (depending on the stimulus, noise, or other factors).

Step 4: Linking to the Labels (yy)

Each Excel file (epoch) corresponds to a single value in yy, which indicates the event type or condition. For example:

Epoch	Label (Event Type)
  1	    1   (Face stimulus)
  2	    2   (Car stimulus)
  3    	1   (Face stimulus)
  …	    …       …
 595   	4   (Scrambled car)

Step 5: Overall Data Structure

Think of your data as 595 Excel files:
•	Each file has 33 rows (EEG channels) and 820 columns (time points).
•	You also have a separate label file (yy) that tells you what condition or event type corresponds to each Excel file.
•	For the table we created for epoch 1 with a 33×820 table, it's name is 1 (Face stimulus) according to yy.
•	x (first dimension): Epochs (each epoch is like a “table”).
•	y (second dimension): Channels (columns in the table).
•	z (third dimension): Time points (rows in the table).

3. Normalization (Z-score normalization)

Normalization is a standard pre-processing step in machine learning and signal processing to ensure that all features (in this case, the EEG signal values across all epochs) are on the same scale. Without normalization, features with larger scales (like EEG channels with higher amplitude signals) might dominate over features with smaller scales (like signals with less variance). This can make the training of machine learning models less effective.

Z-score Normalization is a type of standardization, where we scale the data such that it has a mean of 0 and a standard deviation of 1. 

Z = \frac{X - \mu}{\sigma}

Where:
•	X is the original data point (EEG signal value at a specific time and channel),
•	\mu is the mean of the data (average signal value),
•	\sigma is the standard deviation (how much the data varies from the mean).

This normalization ensures that the data has a consistent scale, which helps improve the performance of machine learning algorithms, especially deep learning models.

4. For example 

x = [
    [1, 2, 3, 4],  # Epoch 1
    [5, 6, 7, 8],  # Epoch 2
    [9, 10, 11, 12] # Epoch 3
]

Here, each row corresponds to an epoch (a trial), and each column corresponds to a feature (in this case, 4 features).
	•	Epoch 1: [1, 2, 3, 4]
	•	Epoch 2: [5, 6, 7, 8]
	•	Epoch 3: [9, 10, 11, 12]

Step-by-Step Z-Score Normalization


Mean and Standard Deviation Calculation:
For each feature (column), we calculate the mean and standard deviation:
	
  •	Mean of Feature 1:  \frac{1 + 5 + 9}{3} = 5 
	•	Mean of Feature 2:  \frac{2 + 6 + 10}{3} = 6 
	•	Mean of Feature 3:  \frac{3 + 7 + 11}{3} = 7 
	•	Mean of Feature 4:  \frac{4 + 8 + 12}{3} = 8 
	
  •	Standard Deviation of Feature 1:
 \sqrt{\frac{(1-5)^2 + (5-5)^2 + (9-5)^2}{3}} = 3.464 
	•	Standard Deviation of Feature 2:
 \sqrt{\frac{(2-6)^2 + (6-6)^2 + (10-6)^2}{3}} = 3.464 
	•	Standard Deviation of Feature 3:
 \sqrt{\frac{(3-7)^2 + (7-7)^2 + (11-7)^2}{3}} = 3.464 
	•	Standard Deviation of Feature 4:
 \sqrt{\frac{(4-8)^2 + (8-8)^2 + (12-8)^2}{3}} = 3.464 

Apply Z-Score Normalization:

Z-score normalization is applied by the formula:

z = \frac{x - \mu}{\sigma}

where:
	•	 x  is the original value,
	•	 \mu  is the mean of the feature,
	•	 \sigma  is the standard deviation of the feature.
For Epoch 1, applying Z-score to each feature:
	•	Feature 1:  \frac{1 - 5}{3.464} = -1.155 
	•	Feature 2:  \frac{2 - 6}{3.464} = -1.155 
	•	Feature 3:  \frac{3 - 7}{3.464} = -1.155 
	•	Feature 4:  \frac{4 - 8}{3.464} = -1.155 
Normalized Epoch 1: [-1.155, -1.155, -1.155, -1.155]

Similarly, apply the Z-score normalization for Epoch 2 and Epoch 3:

Normalized Epoch 2:
	•	Feature 1:  \frac{5 - 5}{3.464} = 0 
	•	Feature 2:  \frac{6 - 6}{3.464} = 0 
	•	Feature 3:  \frac{7 - 7}{3.464} = 0 
	•	Feature 4:  \frac{8 - 8}{3.464} = 0 
Normalized Epoch 2: [0, 0, 0, 0]

Normalized Epoch 3:
	•	Feature 1:  \frac{9 - 5}{3.464} = 1.155 
	•	Feature 2:  \frac{10 - 6}{3.464} = 1.155 
	•	Feature 3:  \frac{11 - 7}{3.464} = 1.155 
	•	Feature 4:  \frac{12 - 8}{3.464} = 1.155 
Normalized Epoch 3: [1.155, 1.155, 1.155, 1.155]

x = [
    [-1.155, -1.155, -1.155, -1.155],  # Epoch 1
    [0,     0,     0,     0    ],     # Epoch 2
    [1.155, 1.155, 1.155, 1.155]      # Epoch 3
]


In [ ]:
# Normalize data per channel (not global)
def normalize_per_channel(X_data):
    # Creates an array of the same shape as X_data, but with all values initialized to zero. 
    # The purpose of this array is to store the normalized data.
    X_norm = np.zeros_like(X_data)
    for i in range(X_data.shape[1]):  # Loop over the range channels: total 4 times from i=0 to i=3
        scaler = StandardScaler()
        # X_channel will have a shape of (6104, 61), representing the data for all trials across all time points for channel i
        # X_data = [[[a1], [a2], [a3], [a4]], 
        #           [[b1], [b2], [b3], [b4]], 
        #           [[c1], [c2], [c3], [c4]]]
        # X_channel = [[a1],[b1],[c1]]
        X_channel = X_data[:, i, :]
        # First, it computes the mean and standard deviation for each feature (each column) in X_channel. 
        # After calculating the mean and standard deviation, it transforms the data by subtracting the mean and dividing by the standard deviation
        X_norm[:, i, :] = scaler.fit_transform(X_channel)
    return X_norm

X_normalized = normalize_per_channel(X_focused)

# Reshape data for Conv2D (samples, channels, times, depth/feature)
X = X_normalized[..., np.newaxis]  # Shape: (samples, channels, times, 1)

'''
Before the np.newaxis operation:
•	The shape of your array X_normalized is something like (samples, channels, time_points).
•	The data is structured in a 3D array with samples as the first dimension, channels as the second dimension, and time points as the third dimension.
•	Each time point for every channel in every sample is a scalar value (real number), and the array looks like this:
 array([[[ 3.24529244e-01,  4.11996053e-01,  4.91609174e-01, ...],
         [ 1.84049422e-01,  2.35332724e-01,  2.83798373e-01, ...],
         [-2.38773378e-01, -2.04526201e-01, -1.74333753e-01, ...],
         [ 4.25627014e-01,  4.73232880e-01,  5.16544162e-01, ...]],

        [[-2.96749702e-01, -2.86137936e-01, -2.64685230e-01, ...],
         [-4.21221455e-01, -4.03617815e-01, -3.79205829e-01, ...],
         [ 2.98252681e-01,  2.99339691e-01,  3.04971293e-01, ...],
         [-2.24049276e-02, -2.27949784e-02, -2.29925722e-02, ...]],

        [[-4.12923850e-01, -3.98015639e-01, -3.82177900e-01, ...], 
        ...

After the np.newaxis operation:
•	The shape changes to (samples, channels, time_points, 1).
•	This operation wraps each individual value (the scalar for each time point) into its own array (of size 1) and adds a new dimension.
•	Now, each time point for each channel/sample is a 1D array (each value in your array is now encapsulated as a single-element list or array).
array([[[[ 3.24529244e-01],
         [ 4.11996053e-01],
         [ 4.91609174e-01],
         ...],
 
        [[ 1.84049422e-01],
         [ 2.35332724e-01],
         [ 2.83798373e-01],
         ...],

        [[-2.38773378e-01],
         [-2.04526201e-01],
         [-1.74333753e-01],
         ...],

        [[ 4.25627014e-01],
         [ 4.73232880e-01],
         [ 5.16544162e-01],
         ...]]], 
       ...
'''

'\nBefore the np.newaxis operation:\n•\tThe shape of your array X_normalized is something like (samples, channels, time_points).\n•\tThe data is structured in a 3D array with samples as the first dimension, channels as the second dimension, and time points as the third dimension.\n•\tEach time point for every channel in every sample is a scalar value (real number), and the array looks like this:\n array([[[ 3.24529244e-01,  4.11996053e-01,  4.91609174e-01, ...],\n        [ 1.84049422e-01,  2.35332724e-01,  2.83798373e-01, ...],\n        [-2.38773378e-01, -2.04526201e-01, -1.74333753e-01, ...],\n        [ 4.25627014e-01,  4.73232880e-01,  5.16544162e-01, ...]],\n\n       [[-2.96749702e-01, -2.86137936e-01, -2.64685230e-01, ...],\n        [-4.21221455e-01, -4.03617815e-01, -3.79205829e-01, ...],\n        [ 2.98252681e-01,  2.99339691e-01,  3.04971293e-01, ...],\n        [-2.24049276e-02, -2.27949784e-02, -2.29925722e-02, ...]],\n\n       [[-4.12923850e-01, -3.98015639e-01, -3.82177900e-0

In [7]:
# Data Augmentation
def augment_data(X, y):
    
    X_augmented = []
    y_augmented = []
    
    # Loop over each sample (trial or instance) in X
    for i in range(X.shape[0]):
        
        # Original data
        # For each sample, we first append the original data (X[i]) and the corresponding label (y[i]) to the X_augmented and y_augmented lists. 
        # This ensures that the original data is kept in the augmented dataset.
        X_augmented.append(X[i])
        y_augmented.append(y[i])

        # Time-shifted data
        # This block generates new samples by shifting the time steps of the data. 
        # The np.roll() function is used to shift the data along the time axis (axis=2). 
        # It shifts the entire time series by 1 or 2 time steps in both forward and backward directions (given by the shift values [-2, -1, 1, 2]).
	    # X_shifted = np.roll(X[i], shift, axis=2) shifts the time data in X[i] by shift units along the time axis.
        '''
        X[0] (shape: (3, 4)):
        [[0.1, 0.2, 0.3, 0.4],
         [0.5, 0.6, 0.7, 0.8],
         [0.9, 1.0, 1.1, 1.2]]
        
        X_shifted = np.roll(X[0], 1, axis=2)
        
        X_shifted (shift = 1):
        [[0.4, 0.1, 0.2, 0.3],
         [0.8, 0.5, 0.6, 0.7],
         [1.2, 0.9, 1.0, 1.1]]
        '''
	    # The shifted data (X_shifted) is then appended to the augmented dataset (X_augmented), and the corresponding label (y[i]) is appended to y_augmented.
        # Shifting the data with np.roll can potentially disrupt the chronological order of the data, and this could negatively affect the ability of the model to learn meaningful patterns from the data, especially when detecting a time-sensitive event like the N170.
        for shift in [-2, -1, 1, 2]:  # Shift by 1 or 2 time steps
            X_shifted = np.roll(X[i], shift, axis = 2)
            X_augmented.append(X_shifted)
            y_augmented.append(y[i])
        
        # Noise-injected data
        # We add noise to the data to create new augmented samples that simulate real-world variations.
	    # noise = np.random.normal(0, 0.01, X[i].shape) generates random noise from a normal distribution with a mean of 0 and a standard deviation of 0.01. The noise has the same shape as the original data sample X[i].
	    # X_noisy = X[i] + noise adds the generated noise to the original sample.
	    # The noisy data (X_noisy) is then appended to X_augmented, and the corresponding label (y[i]) is appended to y_augmented.
        noise = np.random.normal(0, 0.01, X[i].shape)
        X_noisy = X[i] + noise
        X_augmented.append(X_noisy)
        y_augmented.append(y[i])

    return np.array(X_augmented), np.array(y_augmented)

# Apply augmentation on the whole dataset before splitting
X_aug, y_aug = augment_data(X, y)

print(f'Augmented data shape: X={X_aug.shape}, y={y_aug.shape}')



AxisError: axis 2 is out of bounds for array of dimension 2

In [ ]:
# CNN Model Implementation using DeepConvNet
def DeepConvNet(nb_classes, Chans = 4, Samples = 50, dropoutRate = 0.5):
    
    input_main = Input((Chans, Samples, 1))
    
    # Block 1
    block1 = Conv2D(25, (1, 5), padding='same', use_bias=False)(input_main)
    block1 = Conv2D(25, (Chans, 1), use_bias=False)(block1)
    block1 = BatchNormalization()(block1)
    block1 = Activation('elu')(block1)
    block1 = MaxPooling2D((1, 2))(block1)
    block1 = Dropout(0.2)(block1)
    
    # Block 2
    block2 = Conv2D(50, (1, 5), padding='same', use_bias=False)(block1)
    block2 = BatchNormalization()(block2)
    block2 = Activation('elu')(block2)
    block2 = MaxPooling2D((1, 2))(block2)
    block2 = Dropout(0.3)(block2)
    
    # Block 3
    block3 = Conv2D(100, (1, 5), padding='same', use_bias=False)(block2)
    block3 = BatchNormalization()(block3)
    block3 = Activation('elu')(block3)
    block3 = MaxPooling2D((1, 2))(block3)
    block3 = Dropout(0.4)(block3)
    
    # Block 4
    block4 = Conv2D(200, (1, 5), padding='same', use_bias=False)(block3)
    block4 = BatchNormalization()(block4)
    block4 = Activation('elu')(block4)
    block4 = MaxPooling2D((1, 2))(block4)
    block4 = Dropout(0.5)(block4)
    
    # Flatten and Dense Layers
    flatten = Flatten()(block4)
    dense = Dense(nb_classes)(flatten)
    softmax = Activation('softmax')(dense)
    
    return Model(inputs = input_main, outputs = softmax)

# Parameters for DeepConvNet
Chans = X.shape[1]
Samples = X.shape[2]
nb_classes = 2  # Binary classification

# Compile the model
model = DeepConvNet(nb_classes = nb_classes, 
                    Chans = Chans, 
                    Samples = Samples, 
                    dropoutRate = 0.5)

model.compile(optimizer = Adam(learning_rate=1e-3),
              loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy'])
# Adjust EarlyStopping and ReduceLROnPlateau
early_stopping = EarlyStopping(monitor='val_loss', 
                               patience=25, 
                               restore_best_weights=True)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', 
                              factor=0.5, 
                              patience=10,
                              min_lr=1e-6, 
                              verbose=1)

In [ ]:
# Training
epochs = 500
batch_size = 16  # Keep batch size small for better gradient estimation

# Split data into training and testing sets
X_train_val, X_test, y_train_val, y_test = train_test_split(X_aug, 
                                                            y_aug, 
                                                            test_size = 0.2, 
                                                            random_state = 42, 
                                                            stratify = y_aug)

# Train the model
history = model.fit(X_train_val, 
                    y_train_val,
                    epochs = epochs,
                    batch_size = batch_size,
                    validation_data = (X_test, y_test),
                    callbacks = [early_stopping, reduce_lr],
                    verbose = 1)



In [ ]:
# Evaluation and Visualization
# Evaluate on test data
test_loss, test_accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f'Test Accuracy: {test_accuracy * 100:.2f}%')

# Visualize training history
# Plot accuracy
plt.figure(figsize=(12, 4))
plt.subplot(1, 2, 1)
plt.plot(history.history['accuracy'], label='Train Accuracy', color='blue')
plt.plot(history.history['val_accuracy'], label='Validation Accuracy', color='orange')
plt.title('Model Accuracy')
plt.xlabel('Epoch')
plt.ylabel('Accuracy')
plt.legend()

# Plot loss
plt.subplot(1, 2, 2)
plt.plot(history.history['loss'], label='Train Loss', color='blue')
plt.plot(history.history['val_loss'], label='Validation Loss', color='orange')
plt.title('Model Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend()

plt.tight_layout()
plt.show()



In [ ]:
# Predictions
y_pred_prob = model.predict(X_test)
y_pred = np.argmax(y_pred_prob, axis=1)

# Confusion Matrix and Classification Report
cm = confusion_matrix(y_test, y_pred)

plt.figure(figsize=(8, 6))

sns.heatmap(cm, 
            annot=True, 
            fmt='d', 
            cmap='Blues',
            xticklabels=['Face', 'Car'],
            yticklabels=['Face', 'Car'])

plt.xlabel('Predicted Label')

plt.ylabel('True Label')

plt.title('Confusion Matrix')

plt.show()

print(classification_report(y_test, y_pred, target_names=['Face', 'Car']))


In [ ]:
model.save('/Users/owenanderson/Documents/NeurEx/Projects/P1_N170/Models/V2.keras')